Fine grain QueueHints for NodeAffinity plugin

dom4ha opened the pull request:
Skip queue on unrelated change that keeps pod schedulable when QueueHints are enabled.

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
Part of #127405

#### Special notes for your reviewer:

Executed the following tests:

- make test WHAT="pkg/scheduler/..." KUBE_TIMEOUT="-timeout=600s" GOFLAGS="-count=1"
- make test-integration WHAT=./test/integration/scheduler_perf KUBE_TEST_ARGS=-use-testing-log
- make test-integration WHAT=./test/integration/scheduler

#### Does this PR introduce a user-facing change?
```release-note
Improve Node QueueHint in the NodeAffinty plugin by ignoring unrelated changes that keep pods unschedulable.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

dom4ha committed:
Fine grain QueueHints for nodeaffinity plugin.

Skip queue on unrelated change that keeps pod schedulable when QueueHints are enabled.

Split add from QHints disabled case

Remove case when QHints are disabled

Remove two GHint alternatives in unit tests

more fine-grained Node QHint for NodeResourceFit plugin

Return early when updated Node causes unmatch

Revert "more fine-grained Node QHint for NodeResourceFit plugin"

This reverts commit dfbceb60e0c1c4e47748c12722d9ed6dba1a8366.

Add integration test for requeue of a pod previously rejected by NodeAffinity plugin when a suitable Node is added

Add integratin test for a Node update operation that does not trigger requeue in NodeAffinity plugin

Remove innacurrate comment

Apply review comments

sanposhiho commented:
/ok-to-test

/assign @macsko 

for the first path. 
I'll review afterward, for `/approve`.

dom4ha commented:
/cc @alculquicondor

sanposhiho commented:
/release-note-edit
```release-note
Improve Node QueueHint in the NodeAffinty plugin by ignoring unrelated changes that keep pods unschedulable.
```

dom4ha commented:
Is there an annotation that can mark one of the tasks in issues as fixed?

sanposhiho commented:
No, AFAIK. I always manually edit the list 😅

AxeZhan commented on pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:
```diff
@@ -120,13 +120,24 @@ func (pl *NodeAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1
+		if originalNode == nil || !pl.enableSchedulingQueueHint {
+			// The node was created or updated with QueueHints disabled (preCheck could have blocked some previous updates),
+			// so unblocking if only the new Pod is matching.
+			logger.V(4).Info("node was created or updated, and matches with the pod's NodeAffinity", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
```
```suggestion
			logger.V(4).Info("node was created, and matches with the pod's NodeAffinity", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
```
Since we have already checked `originalNode == nil`, we can determine it's a node add event here?

dom4ha replied:
Not in the case when QHints were disabled. Actually I will split it into two separate checks to make the code more readable.

AxeZhan commented on pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:
```diff
@@ -120,13 +120,24 @@ func (pl *NodeAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1
+		if err != nil {
+			return framework.Queue, err
+		}
+		if pl.enableSchedulingQueueHint && isMatched && !wasMatched {
```
```suggestion
		if isMatched && !wasMatched {
```
same here

macsko replied:
Even `if !wasMatched` is enough, as `isMatched` is already checked above.

dom4ha replied:
True

macsko commented on pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go:
```diff
@@ -1234,8 +1234,11 @@ func TestNodeAffinityPriority(t *testing.T) {
+	qhintDisabled := false
 	podWithNodeAffinity := st.MakePod().NodeAffinityIn("foo", []string{"bar"}, st.NodeSelectorTypeMatchExpressions)
 	testcases := map[string]struct {
+		qhints         *bool
```
You could try to make it similar to other tests (see [here](https://github.com/kubernetes/kubernetes/blob/8adc35ebd7221683b9f9ec7a9e2c63cf3f6d0dff/test/integration/scheduler/queue_test.go#L204-L206)), so we won't need `qhintEnabled` and `qhintDisabled` variables then.

dom4ha commented on pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:
```diff
@@ -120,13 +120,24 @@ func (pl *NodeAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1
 	if isMatched {
-		logger.V(4).Info("node was created or updated, and matches with the pod's NodeAffinity", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
-		return framework.Queue, nil
+		if originalNode == nil || !pl.enableSchedulingQueueHint {
```
In my understanding, it will be called for both cases, so we should preserve the existing functionality.

macsko replied:
If queueing hints feature is disabled, hint function is overwritten to defaultQueueingHintFn ([ref](https://github.com/kubernetes/kubernetes/blob/f0415a65b41e5eecc04425992c612832b27915f9/pkg/scheduler/scheduler.go#L415-L417)). Therefore `isSchedulableAfterNodeChange` is called only if queueing hints are enabled.

AxeZhan replied:
I think we'll use `defaultQueueingHintFn` instead when feature is disabled?🤔

dom4ha replied:
Ah, you mena that isSchedulableAfterNodeChange is a feature of QHints, so never call when they are disabled. Make sense.

alculquicondor commented:
LGTM, but I'll leave it to already involved reviewers to give the official one

sanposhiho commented on pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:
```diff
@@ -120,13 +120,22 @@ func (pl *NodeAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1
-		logger.V(4).Info("node was created or updated, and matches with the pod's NodeAffinity", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
-		return framework.Queue, nil
+		if originalNode == nil {
+			logger.V(4).Info("node was created, and matches with the pod's NodeAffinity", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
```
I know it's original issue, but please modify the log level to 5

sanposhiho commented on pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:
```diff
@@ -120,13 +120,22 @@ func (pl *NodeAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1
+			return framework.Queue, err
+		}
+		if !wasMatched {
+			logger.V(4).Info("node was updated and the pod's NodeAffinity changed to matched", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
```
ditto

sanposhiho commented on pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:
```diff
@@ -120,13 +120,22 @@ func (pl *NodeAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1
 		return framework.Queue, err
 	}
 	if isMatched {
```
can we early return `if !isMatched` instead of nesting `if`s?

sanposhiho reviewed (commented):
only nits. overall looks great.

sanposhiho reviewed (commented):
Also, can you add integration test cases for Node/Add and Node/UpdateLabel at -
https://github.com/kubernetes/kubernetes/blob/c348d099f8a2a83ac3a4d8c533245ac8d3a2f6fe/test/integration/scheduler/queue_test.go#L260

Looks like I overlooked they were missing when reviewing [the PR](https://github.com/kubernetes/kubernetes/pull/127139/).

dom4ha commented:
> Also, can you add integration test cases for Node/Add and Node/UpdateLabel at -

Added test case for Node/Add, but I believe that Node/UpdateLabel already exists as "Pod rejected by the NodeAffinity plugin is requeued when a Node is updated"

Regarding testing the functional changes in this PR in integration test, I'm not sure yet how to test them, since the test TestCoreResourceEnqueue is verifying requeueing, but we added a case where we want to prevent from requeueing. I'm looking for options, but was it actually the intention of your comment?

dom4ha commented:
I think the comment https://github.com/kubernetes/kubernetes/pull/127473/files#r1767772258 actually answers my question. Plugins register if they have rejected any node, not this particular one.

Well, theoretically plugins could register per node rejected I guess, so it would simplify further qhints analysis, but I still may miss something.

AxeZhan commented:
> Regarding testing the functional changes in this PR in integration test, I'm not sure yet how to test them, since the test TestCoreResourceEnqueue is verifying requeueing, but we added a case where we want to prevent from requeueing. I'm looking for options, but was it actually the intention of your comment?

You can verify a pod is not requeued with setting a nil `wantRequeuedPods`,
maybe test case `Updating pod condition doesn't retry scheduling if the Pod was rejected by TaintToleration` can help you.

And then maybe we can design this case to verify this pr:

1. Pod got rejected by nodeAffinity on `node-1`, and rejected by noderesource on `node-2`
2. `node-2` updated its label, but since it met the nodeAffinity before, thus the pod isn't requeued.

sanposhiho commented:
> Plugins register if they have rejected any node, not this particular one.
Well, theoretically plugins could register per node rejected I guess, so it would simplify further qhints analysis

We cannot do this simply.
The point is that, let's say, even if plugin-A rejects only Node-A, the plugin-A could change the scheduling result with Node/Update to other nodes. I call such plugins cross-node scheduling constraints, personally; In in-tree plugins, PodTopologySpread is the only one. (and, ofc, there could be a lot in out-of-tree plugin implementations)
So, regarding normal plugins, they just check one Node and decide whether to reject or not; for example, like NodeAffinity here, NodeAffinity plugin checks one Node to see if it matches the affinity or not, and then accepts/rejects the Node. 
But, PodTopologySpread checks the whole situation of each domain and rejects all Nodes in the domain.
That's the difference between normal plugins and cross-node scheduling constraint plugins.

As an actual example, let's say pods are spreading 1:2:2 (domain-1/domain-2/domain-3) to three domains and the pods have `maxSkew:1`, a new pod can only go to domain-1. topology spread plugin rejects all nodes in domain-2 and domain-3.
And, then, say somehow all nodes in domain1 change their `topologyKey` labels and join domain2; It changes the whole spreading situation from 1:2:2 to 3:2 (b/c topology spread cannot see domain-1 anymore) and now the plugin allows the Pod to go to domain-3. Meaning, the label updates to the nodes in domain-1 (which weren't rejected by topology spread) could indirectly make the Pod schedulable in nodes in another domain.

---

Actually, I have a dream future idea in QHint, especially around when to requeue Pods and how to trigger QHints. I'm pretty sure we definitely can have a further improvement there.
But, in any case, it's not that simple and we shouldn't do it right now; we should aim to make the current simple design of QHint stable first.

sanposhiho commented on test/integration/scheduler/queue_test.go:
```diff
@@ -277,6 +277,26 @@ func TestCoreResourceEnqueue(t *testing.T) {
 			},
 			wantRequeuedPods: sets.New("pod1"),
 		},
```
Can you improve `Pod rejected by the NodeAffinity plugin is requeued when a Node is updated` to see if match -> match update is ignored correctly, which makes this PR's benefit visible?

AxeZhan commented on test/integration/scheduler/queue_test.go:
```diff
@@ -271,7 +271,48 @@ func TestCoreResourceEnqueue(t *testing.T) {
+			triggerFn: func(testCtx *testutils.TestContext) error {
+				// Trigger a NodeUpdate event to add new label.
+				// It won't cause pod to be requeued, because there was a match already before the update, meaning this plugin wasn't blocking the scheduling.
+				if _, err := testCtx.ClientSet.CoreV1().Nodes().Update(testCtx.Ctx, st.MakeNode().Name("node1").Label("group", "a").Label("group", "b").Obj(), metav1.UpdateOptions{}); err != nil {
```
Can a node has two label with the same key?
Won't this results in the node has the label `group:b`, and thus we returned in `isSchedulableAfterNodeChange` L122?

alculquicondor commented on pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:
```diff
@@ -119,16 +119,27 @@ func (pl *NodeAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1
-	if isMatched {
-		logger.V(4).Info("node was created or updated, and matches with the pod's NodeAffinity", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+	if !isMatched {
+		logger.V(5).Info("node was created or updated, but the pod's NodeAffinity is unmatch", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
```
```suggestion
		logger.V(5).Info("node was created or updated, but the pod's NodeAffinity doesn't match", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
```

AxeZhan commented:
/lgtm

Can you squash the commits though.

dom4ha commented:
> Can you squash the commits though.

I hope I did it right :)

AxeZhan commented:
> I hope I did it right :)

Yes :) 
/lgtm

alculquicondor commented:
/retest

alculquicondor commented:
/approve
